Give game makers more control

phrohdoh opened the pull request:
This adds to the `Info.plist`:
 - The string `{BUNDLE_DISPLAY_NAME}` to be replaced (via `sed`, or `plutil`, etc) by the game's title (this could be renamed `GAME_DISPLAY_NAME` or even `GAME_NAME` if desired)
 - The string `{DEFAULT_MOD_ID}` which would (as an example for RA2) be replaced with `ra2` (this defaults to ~~`modchooser`~~ `nil` and if not supplied `Game.Mod` and `Engine.DefaultMod` will not be passed)

Changes to the build system:
 - Moved output assemblies into a `build` directory which the `clean` target removes

phrohdoh committed:
Improve makefile behaviour:

* Output build products in build subdirectory
* Suppress unnecessary terminal output
* Fix 'clean' Makefile target

pchote commented on OpenRA.m:
```diff
@@ -121,15 +121,20 @@ - (void)launchGameWithArgs: (NSArray *)gameArgs
-	// Default values - can be overriden by setting MonoMinVersion and MonoGameExe in Info.plist
+	// Default values - can be overriden by setting certain keys Info.plist
 	NSString *gameName = @"OpenRA.Game.exe";
+	NSString *defaultModId = @"modchooser";
```
Hardcoding this to modchooser isn't great considering that we're about to kill that.  Can you please default this to `nil`, and then only add the launch arguments if it is non-nil?

phrohdoh replied:
Done.

phrohdoh commented:
Updated.

pchote commented on Makefile:
```diff
@@ -1,22 +1,28 @@
 launchgame: launchgame.m
-	clang -m32 launchgame.m -o launchgame -framework AppKit -mmacosx-version-min=10.6
+	@mkdir -p build
+	@clang -m32 launchgame.m -o build/launchgame -framework AppKit -mmacosx-version-min=10.6
```
I'm not a fan of suppressing *all* output from the makefile steps.  Can you please print out "Compiling launchgame", "Compiling OpenRA", "Compressing launcher.zip" etc?

pchote commented on OpenRA.m:
```diff
@@ -121,15 +121,20 @@ - (void)launchGameWithArgs: (NSArray *)gameArgs
 		if (exeValue && [exeValue length] > 0)
 			gameName = exeValue;
+
+		NSString *defaultModIdValue = [plist objectForKey:@"DefaultModId"];
```
I've come to the conclusion that our only feasible way forwards on the `OpenRA.Game.exe` side is to make `Game.Mod` a required commandline parameter and remove the problematic `Mod` settings.yaml key that is causing all our problems.  Once that is done `Engine.DefaultMod` will be redundant and can be removed.

Looking ahead to this change, can you please drop the "default" from the plist key and variable names?

pchote commented on OpenRA.m:
```diff
@@ -142,6 +147,13 @@ - (void)launchGameWithArgs: (NSArray *)gameArgs
+
+	if (defaultModId)
+	{
+		[launchArgs addObject: [NSString stringWithFormat:@"Engine.DefaultMod=%@", defaultModId]];
```
We can keep this for now, because it will be needed during the transition and won't do any harm after it is complete.

pchote commented:
Can you please also extract the hardcoded FAQ link at [OpenRA.m#L59](https://github.com/OpenRA/OpenRALauncherOSX/blob/master/OpenRA.m#L59) to an info.plist key?  Third party mods may want to change this to.

phrohdoh commented on Info.plist:
```diff
@@ -119,5 +119,9 @@
 			</dict>
 		</dict>
 	</array>
+	<key>DefaultModId</key>
```
I need to rename this too.

pchote committed:
Remove oramod and orarep file handlers.

pchote committed:
Allow mods to customize OSX integration:

* {MOD_NAME}: Title (< 16 characters) shown in
  menu bar, about panel, and error dialogs.
* {MOD_ID}: Mod id given to the OpenRA engine.
* {JOIN_SERVER_URL_SCHEME}: The URL scheme that
  can be used to directly join a server.
* {FAQ_URL}: URL to open from the FAQ button in
  the crash dialog.

pchote committed:
Remove RA mod icon.

pchote commented:
The launcher needs several additional changes to support the things that I want for our split official mod launchers.  Rather than get you to write what I have in mind by proxy I have taken this as a base, and been working on it at [per-mod-template](https://github.com/OpenRA/OpenRALauncherOSX/compare/per-mod-template?expand=1) in parallel with OpenRA/OpenRA packaging script changes.

If you're up for it I'll finish my changes, add back your makefile changes (plus my review requests), and then push it over this PR for review and merge.

pchote committed:
Fix a memory leak in [getUrl: withReplyEvent:]

phrohdoh commented:
Sounds good to me.

pchote commented:
Ok, done.  This gets a :+1: from me as-is.  If you're happy with my changes then :+1: and merge.

phrohdoh commented on Info.plist:
```diff
@@ -5,17 +5,17 @@
 	<key>CFBundleIconFile</key>
 	<string>OpenRA.icns</string>
 	<key>CFBundleIdentifier</key>
-	<string>net.openra.launcher</string>
```
People will probably want to change this entire value but for the sake of moving this work forward we can do that later.
